### DISCUSSION TITLE
Chrome 85 Breaks Referer

### DISCUSSION THREAD

**Discussion by jamesward** [community/community#24373] [Apps, API and Webhooks]
<p>Before Chrome 85, when clicking external links on github the browser would send the <code>Referer</code> header with the full path.  For example, on <a href="https://github.com/jamesward/hello-netcat" rel="nofollow noopener">https://github.com/jamesward/hello-netcat</a> clicking the button goes to a site that uses the <code>Referer</code> header to know which repo the user is coming from.  Cloud Run Button, Heroku Button, and other tools use this header.  As of Chrome 85, the browser sets  <code>Referrer-Policy</code><br>
to <code>strict-origin-when-cross-origin</code> which means these links now only send a <code>Referer</code> of <code>https://github.com</code> which breaks anything that depends on knowing which repo the user was on when they clicked something. Details: <a href="https://developers.google.com/web/updates/2020/07/referrer-policy-new-chrome-default" rel="nofollow noopener">https://developers.google.com/web/updates/2020/07/referrer-policy-new-chrome-default</a></p>
<p>It appears that <code>github.com</code> can override this default by sending their own <code>Referrer-Policy</code> header with a value of <code>no-referrer-when-downgrade</code>.  Would it be possible to do this and unbreak a bunch of stuff?</p>

**Comment by maudnals** [comment]
<p>Hi there,<br>
Chrome team involved with the change in question here (<em>maudnals</em>, <em>cos_theta</em>, <em>rowan_m</em> on Twitter). A few ideas to share, we hope this may help!</p>
<p><a class="mention" href="/jamesward">@jamesward</a> 's suggestion would fix the problem, but full GitHub URLs would always be leaked in all outgoing cross-origin requests across all of GitHub, which hinders users privacy. Browsers are generally <a href="https://web.dev/referrer-best-practices/#default-referrer-policies-in-browsers" rel="nofollow noopener">switching to <code>strict-origin-when-cross-origin</code></a> as a default for this reason. This also may not be future-proof, in case browsers become more strict and stop sharing full URLs in <code>Referer</code>s in cross-origin requests.</p>
<hr>
<p>A few ideas/solutions:</p>
<p><strong>Solution 1</strong><br>
[EDIT: not working]<br>
Button users now must include HTML to specify a referrer policy instead of simple markdown. <code>&lt;a referrerpolicy="no-referrer-when-downgrade" href="https://deploy.cloud.run"&gt;&lt;img src="https://deploy.cloud.run/button.svg" alt="Run on Google Cloud"&gt;&lt;/a&gt;</code>.<br>
=&gt; Example here: <a href="https://github.com/maudnals/button-sample/blob/master/README.md" rel="nofollow noopener">https://github.com/maudnals/button-sample/blob/master/README.md</a></p>
<p>Pros:</p>
<ul>
<li>GH users’ privacy is preserved</li>
<li>Doable “today/soon”? (would require Heroku/Cloud Run documentation change)</li>
</ul>
<p>Cons:</p>
<ul>
<li>May not be future-proof, in case browsers become more strict and stop sharing full URLs in <code>Referer</code> cross-origin.</li>
<li>May not be supported on Safari <strong>iOS</strong>.</li>
</ul>
<p>(A variation of this would be update the markdown to support specifying the <code>referrerpolicy</code> attribute.)</p>
<p><strong>Solution 2</strong><br>
Button users (= people that include a Heroku/Cloud Run button on their GH repo) now <em>always</em> need to specify the GH repo. Heroku/Cloud Run scripts never rely on the <code>Referer</code> anymore.</p>
<p>Pros:</p>
<ul>
<li>GH users’ privacy is preserved.</li>
</ul>
<p>Cons:</p>
<ul>
<li>Heroku/Cloud Run/Other tools need to update their systems to not expect a <code>Referer</code>.</li>
<li>For GH users, forking repos that include a Heroku/Cloud Run button requires one more step: changing the repo URL in the Button configuration. This may lead to errors if they forget to do so. Mitigation: Heroku/Cloud Run may update their documentation to include a warning: "When including the button on a GH repo, add a line on your README that says: ‘Change this if you fork this repo!’ ".</li>
<li>Potentially more future-proof. Not dependent on browser behaviour.</li>
</ul>
<hr>
<p>Side note for GitHub: regardless of the solution, applying an explicit privacy-preserving policy of <code>strict-origin-when-cross-origin</code> globally to <code>github.com</code> would help it behave predictably. Because if <code>github.com</code> has no policy at all, it will fall back to the browser default policy i.e. <code>strict-origin-when-cross-origin</code> or other, depending on the browser - the behaviour won’t be consistent across browsers and will evolve as browser default changes.</p>
<p><em>maudnals</em>, <em>cos_theta</em>, <em>rowan_m</em> on Twitter</p>

**Comment by maudnals** [comment]
<p>(<a href="https://caniuse.com/mdn-html_elements_a_referrerpolicy" rel="nofollow noopener">referrerpolicy on Safari iOS</a> - Not tested though!)</p>

**Comment by jamesward** [comment]
<p>Thanks for the response and investigation!  It doesn’t look like the sample is working.  Shouldn’t the <code>referrerpolicy="no-referrer-when-downgrade"</code> be on the <code>a</code> tag?  Or maybe github is filtering it out in their rendering?</p>

**Comment by airtower-luna** [comment]
<p>Looks to me like the <code>referrerpolicy</code> attribute in the example slipped into the <code>img</code> tag when it should be in the <code>a</code>.</p>

**Comment by maudnals** [comment]
<p>Apologies, it’s now in the <code>&lt;a&gt;</code> but it looks indeed like it’s removed, so it’s not working.</p>
<p>[EDIT: adding a comment by editing this answer, since I’m not allowed to post over 3 - this comment comes after <a class="mention" href="/ahmetb">@ahmetb</a> 's comment below]:</p>
<p>Summing up a few ideas/solutions that GitHub would need to act on:</p>
<ul>
<li>
<p>For all rendered markdown files, set the policy on all <code>&lt;a&gt;</code> elements automatically to <code>no-referrer-when-downgrade</code> (quoting <a class="mention" href="/ahmetb">@ahmetb</a> 's answer from below)</p>
</li>
<li>
<p>Use heuristics to detect Cloud/Heroku buttons and set the policy only on these <code>&lt;a&gt;</code> elements to <code>no-referrer-when-downgrade</code></p>
</li>
<li>
<p>GH would enable <code>referrerpolicy</code> to be set in <code>&lt;a&gt;s</code> in markdown files (not possible now), run a query on their end, and could potentially notify the repository owners to consider updating their code.</p>
</li>
</ul>
<p>Note: any of these would be preferable privacy-wise to setting a global behavior site-wide with a header, as they would limit cross-origin leaks to a smaller surface of GH.</p>
<p>About the privacy bit:</p>
<ul>
<li>
<p>A GitHub repo’s URL contains usernames, which looks like it’s considered <a href="https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/what-is-personal-data/what-are-identifiers-and-related-factors/#pd3:~:text=An%20individual%E2%80%99s%20social%20media%20%E2%80%98handle%E2%80%99%20or,with%20a%20%E2%80%98real%20world%E2%80%99%20named%20individual." rel="nofollow noopener">personal data under the GDPR</a></p>
</li>
<li>
<p>The privacy problem here is primarily about leaking information cross-origins the user may not expect.</p>
</li>
</ul>
<p><em>maudnals</em> ,  <em>cos_theta</em> ,  <em>rowan_m</em>  on Twitter</p>

**Comment by ahmetb** [comment]
<p>At this point, just Google’s button already exists on some triple-digit repos. Those will be broken forever unless we go submit PRs for them manually. Not to mention Azure Websites, Heroku etc have been around longer, so we’re likely looking at thousands of broken repos here.</p>
<p>Feasible solution requested from GitHub here would look like:</p>
<ul>
<li>restore old behavior site-wide with a header, OR</li>
<li>add <code>referrerpolicy</code> property to <code>&lt;a&gt;</code> elements, for either:
<ul>
<li>in all rendered markdown files, OR</li>
<li>whitelist certain button urls (just like how certain badge images are freshly loaded in .md files on github)</li>
</ul>
</li>
</ul>
<p>I’m inclined to think there aren’t privacy issues with this as far as GitHub users are concerned. The repos in question are public repos and don’t contain PII.</p>

**Comment by jamesward** [comment, marked as answer]
<p>It appears that this has been fixed. :clap:</p>

**Reply by hirehamir** [reply]
I still see an error at https://deploy.cloud.run/ that links to this page.

**Reply by hirehamir** [reply]
I got there from the `Run on Google Cloud` button at https://github.com/mixpanel/tracking-proxy.
